kube-derive/src/custom_resource.rs

```diff
@@ -64,6 +66,42 @@ fn default_serde_json() -> Path {
     parse_quote! { ::serde_json }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum SchemaMode {
+    Disabled,
+    Custom,
+    Derived,
```

Comment:
I originally misunderstood the intention of these names, so proposing some new ones:

- `Disabled` -> `Elide` (to distance ourselves from the separate requirement of having these)
- `Custom` -> `Specialize` (to indicate more clearn that a user has to specialize its impl)
- `Derived` -> `Derive` ~~or `Auto`~~ (to more indicate no user action needed)

That also potentially reads nicer with a short-hand name:

`#[kube(schema = "Elide")]`

Reply:
although, that said, the imperative form of the verbs, while helpful for `Derive` ("you do it program"), is probably not so helpful for `Specialize`..

Reply:
maybe we just have two bools instead:

- `#[kube(schema_specialize)`
- `#[kube(schema_elide)`

that matches the internal model (and the bools have no meaningful overlap)

Reply:
I'm not married to any of the current names, but I'm not sold on any of your proposed changes either.

> Disabled -> Elide (to distance ourselves from the separate requirement of having these)

`Elide` sounds like [lifetime elision](https://doc.rust-lang.org/reference/lifetime-elision.html) to me (in a Rust context).. the lifetimes still exist but we can imply them from the context (which, if anything, is closer to deriving than disabling IMO)

> Custom -> Specialize (to indicate more clearn that a user has to specialize its impl)

This sounds like trait/type specialization, and I think the connection might confuse more than it helps..

> Derived -> Derive or Auto (to more indicate no user action needed)

`Auto` implies to me that we automatically select the mode (which is currently implied when there is no `schema_mode` attribute)

> two bools

Bleh.

---

That said, I think this raises the question about whether `SchemaMode::Disabled` is operating at the wrong abstraction level.. Maybe the better question would be whether we should have a flag to disable `Self::crd()` wholesale (for a given kind)?

Reply:
On the other hand, it might be worth keeping `SchemaMode::Disabled` around (in some form) as a migration path for people who have used the current override...

Reply:
Taking a synonyms/antonym approach here to avoid potential confusion with rust feature names:

- instead of `Disabled` / `Elide` we could use `Omit` - for Omit impl
- instead of `Custom` / `Specialize` we could use `Manual` - for Manual impl

One problem here i think is that the third option (Derive default) is a bit unnecessary to have (and it's hard to give it an antonym name).

Maybe it would make sense to detect `manual` or `derive` mode based on whether people `#[derive(JsonSchema)]` on the spec struct?

Reply:
> Maybe it would make sense to detect manual or derive mode based on whether people #[derive(JsonSchema)] on the spec struct?

That was my hope, but it seems like the `#[derive]` is stripped before the token stream is passed to the macro.

To be honest, I would have preferred to not special-case the derive at all, and just let people `#[kube(derive(JsonSchema))]` it if they wanted to.

But that ship has long sailed.

Reply:
> That was my hope, but it seems like the `#[derive]` is stripped before the token stream is passed to the macro.

What happens if you add `forward_attrs(derive)`?
```rust
#[derive(Debug, FromDeriveInput)]
#[darling(attributes(kube), forward_attrs(derive))]
struct KubeAttrs {
```
Is `derive_input.attrs` still empty? https://docs.rs/darling/0.13.0/darling/#fromderiveinput

Reply:
Yeah, you can dump the full input in `kube-derive/lib.rs` and the doc comments before are in fact available to the input, but not the root `#[derive(..)]` :-(

Reply:
I see `derive` if I move them into a separate `derive`, but that's useless:
```rust
#[derive(CustomResource)]
//
#[derive(Serialize, Deserialize, Debug, PartialEq, Clone, JsonSchema)]
struct ExampleSpec {
}
```

Reply:
Yeah, we can't do this. https://github.com/rust-lang/rust/issues/40574#issuecomment-287300647

> To be honest, I would have preferred to not special-case the derive at all, and just let people `#[kube(derive(JsonSchema))]` it if they wanted to.

How does this work? Maybe I'm misremembering, but I thought we couldn't modify the input struct. Is it possible to add `derive(JsonSchema)` to it? Doesn't the spec struct need `derive(JsonSchema)` to have schema generated?

Reply:
> How does this work?

We cannot modify the input, but we were just thinking about reading that derive property.

The key struct here isn't the spec struct, but the generated root struct, which needs to implement `JsonSchema` (with our injected derive on it, or manually by the user) for the schema to get included in the `::crd()` method.

The only remaining possibility here is that we __could__ use `#[kube(derive = "JsonSchema")]` as an  indicator of whether to automatically derive `JsonSchema` on the root struct (rather than select it as part of a tri-state enum, or, do as we do know inject it by default if the feature is on). But that would be a pretty big breaking change.